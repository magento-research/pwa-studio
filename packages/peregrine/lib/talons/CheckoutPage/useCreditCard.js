import { useCallback, useState } from 'react';

export const useCreditCard = props => {
    const { setDoneEditing } = props;
    const [paymentNonce, setPaymentNonce] = useState();
    const onPaymentSuccess = useCallback(
        nonce => {
            console.log('Payment Nonce Received', nonce);
            setPaymentNonce(nonce);
            setDoneEditing(true);
        },
        [setDoneEditing]
    );
    const onPaymentError = useCallback(error => {
        console.error(error);
    }, []);
    const onPaymentReady = useCallback(data => {
        console.log('payment Ready', data);
    }, []);

    return {
        paymentNonce,
        onPaymentError,
        onPaymentSuccess,
        onPaymentReady
    };
};
