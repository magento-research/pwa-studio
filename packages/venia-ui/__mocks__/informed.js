import React from 'react';

const informed = jest.requireActual('informed');
const {
    Text,
    Input,
    Radio,
    TextArea,
    Select,
    Option,
    Checkbox,
    RadioGroup,
    BasicText,
    BasicRadio,
    BasicRadioGroup,
    BasicTextArea,
    BasicSelect,
    BasicCheckbox,
    Scope,
    ArrayField,
    asField,
    ...rest
} = informed;

const wrapWithUuidPropCheck = Component => props => {
    const newProps = { ...props };
    const uuid4Regex = new RegExp(
        /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i
    );
    const hasUuidIdProp =
        props.hasOwnProperty('id') && uuid4Regex.test(props.id);

    if (!props.hasOwnProperty('id') || hasUuidIdProp) {
        if (props.hasOwnProperty('field')) {
            newProps.id = props.field;
        } else {
            newProps.id = 'SAMPLE-ID';
        }
    }

    return <Component {...newProps} />;
};

module.exports = {
    ...rest,
    Text: wrapWithUuidPropCheck(Text),
    Input: wrapWithUuidPropCheck(Input),
    Radio: wrapWithUuidPropCheck(Radio),
    TextArea: wrapWithUuidPropCheck(TextArea),
    Select: wrapWithUuidPropCheck(Select),
    Option: wrapWithUuidPropCheck(Option),
    Checkbox: wrapWithUuidPropCheck(Checkbox),
    RadioGroup: wrapWithUuidPropCheck(RadioGroup),
    BasicText: wrapWithUuidPropCheck(BasicText),
    BasicRadio: wrapWithUuidPropCheck(BasicRadio),
    BasicRadioGroup: wrapWithUuidPropCheck(BasicRadioGroup),
    BasicTextArea: wrapWithUuidPropCheck(BasicTextArea),
    BasicSelect: wrapWithUuidPropCheck(BasicSelect),
    BasicCheckbox: wrapWithUuidPropCheck(BasicCheckbox),
    Scope: wrapWithUuidPropCheck(Scope),
    ArrayField: wrapWithUuidPropCheck(ArrayField),
    asField: Component => {
        const WrappedComponent = asField(Component);

        return props => {
            const newProps = { ...props };
            const uuid4Regex = new RegExp(
                /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i
            );
            const hasUuidIdProp =
                props.hasOwnProperty('id') && uuid4Regex.test(props.id);

            if (!props.hasOwnProperty('id') || hasUuidIdProp) {
                if (props.hasOwnProperty('field')) {
                    newProps.id = props.field;
                } else {
                    newProps.id = 'SAMPLE-ID';
                }
            }

            return <WrappedComponent {...newProps} />;
        };
    }
};
