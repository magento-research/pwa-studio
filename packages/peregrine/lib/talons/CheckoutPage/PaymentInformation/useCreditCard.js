import { useCallback, useEffect, useState } from 'react';
import { useFormState, useFormApi } from 'informed';
import { useQuery, useApolloClient, useMutation } from '@apollo/react-hooks';

import { useCartContext } from '../../../context/cart';

const mapBillingAddressData = rawBillingAddressData => {
    if (rawBillingAddressData) {
        const { street, country, region } = rawBillingAddressData;

        return {
            ...rawBillingAddressData,
            street1: street[0],
            street2: street[1],
            country: country.code,
            state: region.code
        };
    } else {
        return {};
    }
};

/**
 * Talon to handle Credit Card payment method.
 *
 * @param {Boolean} props.isHidden boolean value which represents if the component is hidden or not
 * @param {Function} props.onSuccess callback to invoke when the a payment nonce has been generated
 * @param {Function} props.onReady callback to invoke when the braintree dropin component is ready
 * @param {Function} props.onError callback to invoke when the braintree dropin component throws an error
 * @param {DocumentNode} props.queries.getAllCountriesQuery query to fetch all countries data
 * @param {DocumentNode} props.queries.getBillingAddressQuery query to fetch billing address from cache
 * @param {DocumentNode} props.queries.getIsBillingAddressSameQuery query to fetch is billing address same checkbox value from cache
 * @param {DocumentNode} props.queries.getPaymentNonceQuery query to fetch payment nonce saved in cache
 *
 * @returns {
 *   onPaymentError: Function,
 *   onPaymentSuccess: Function,
 *   onPaymentReady: Function,
 *   isBillingAddressSame: Boolean,
 *   countries: Object,
 *   isDropinLoading: Boolean
 * }
 */
export const useCreditCard = props => {
    const { onSuccess, queries, mutations, isHidden, onReady, onError } = props;
    const {
        getAllCountriesQuery,
        getBillingAddressQuery,
        getIsBillingAddressSameQuery,
        getPaymentNonceQuery
    } = queries;
    const { setBillingAddressMutation, setCreditCardDetailsOnCart } = mutations;

    /**
     * Definitions
     */

    const [cacheDataRestored, setCacheDataRestored] = useState(false);
    const [isDropinLoading, setDropinLoading] = useState(true);

    const client = useApolloClient();
    const formState = useFormState();
    const formApi = useFormApi();
    const [{ cartId }] = useCartContext();

    const { data: countriesData } = useQuery(getAllCountriesQuery);
    const { data: billingAddressData } = useQuery(getBillingAddressQuery, {
        variables: { cartId }
    });
    const { data: isBillingAddressSameData } = useQuery(
        getIsBillingAddressSameQuery,
        { variables: { cartId } }
    );
    const [updateBillingAddress] = useMutation(setBillingAddressMutation);
    const [updateCCDetails] = useMutation(setCreditCardDetailsOnCart);

    const { countries } = countriesData || {};
    const isBillingAddressSame = formState.values.isBillingAddressSame;

    /**
     * Effects
     */

    useEffect(() => {
        /**
         * If credit card component is hidden, reset
         * `cacheDataRestored` to `false` so when the
         * component is selected, data from cache will
         * be used to restore form state.
         *
         * Similarly reset `isDropinLoading` to `true`
         * so when the component is selected, till the
         * braintree drop in ready, a loading state will
         * be shown.
         */
        if (isHidden) {
            setCacheDataRestored(false);
            setDropinLoading(true);
        }
    }, [isHidden]);

    useEffect(() => {
        /**
         * Perform UI restoration only if all of the below are true
         * 1. Credit card component is not hidden
         * 2. Brain tree drop in is not loading
         * 3. It is the first time
         */
        if (!isHidden && !isDropinLoading && !cacheDataRestored) {
            /**
             * Setting the checkbox to the value in cache
             */
            const isBillingAddressSame = isBillingAddressSameData
                ? isBillingAddressSameData.cart.isBillingAddressSame
                : true;

            formApi.setValue('isBillingAddressSame', isBillingAddressSame);

            /**
             * Setting billing address data from cache
             */
            if (billingAddressData) {
                if (billingAddressData.cart.billingAddress) {
                    const billingAddress = mapBillingAddressData(
                        billingAddressData.cart.billingAddress
                    );
                    // eslint-disable-next-line no-unused-vars
                    const { __typename, ...rest } = billingAddress;
                    formApi.setValues(rest);
                }
            }

            /**
             * Setting so this effect will not be applied on every render
             */
            setCacheDataRestored(true);
        }
    }, [
        cacheDataRestored,
        isBillingAddressSameData,
        formApi,
        billingAddressData,
        isDropinLoading,
        isHidden
    ]);

    /**
     * Helpers
     */

    const setIsBillingAddressSame = useCallback(() => {
        client.writeQuery({
            query: getIsBillingAddressSameQuery,
            data: {
                cart: {
                    __typename: 'Cart',
                    id: cartId,
                    isBillingAddressSame
                }
            }
        });
    }, [client, cartId, getIsBillingAddressSameQuery, isBillingAddressSame]);

    const setBillingAddress = useCallback(() => {
        const {
            firstName = '',
            lastName = '',
            country = '',
            street1 = '',
            street2 = '',
            city = '',
            state = '',
            postalCode = '',
            phoneNumber = ''
        } = formState.values;

        updateBillingAddress({
            variables: {
                cartId,
                firstName,
                lastName,
                street1,
                street2,
                city,
                state,
                postalCode,
                country,
                phoneNumber
            }
        });
    }, [formState.values, updateBillingAddress, cartId]);

    const setPaymentNonce = useCallback(
        paymentNonce => {
            client.writeQuery({
                query: getPaymentNonceQuery,
                data: {
                    cart: {
                        __typename: 'Cart',
                        id: cartId,
                        paymentNonce
                    }
                }
            });
        },
        [cartId, client, getPaymentNonceQuery]
    );

    const updateCCDetailsOnCart = useCallback(
        braintreeNonce => {
            const { nonce } = braintreeNonce;
            updateCCDetails({
                variables: {
                    cartId,
                    paymentMethod: 'braintree',
                    paymentNonce: nonce
                }
            });
        },
        [updateCCDetails, cartId]
    );

    const onPaymentSuccess = useCallback(
        nonce => {
            /**
             * TODO
             * Ideally if the billing address is same as
             * shipping address, we have to fetch shipping
             * address and use it to save billing address.
             *
             * But I have some doubts about shipping addresses.
             * Till those get rectified, ill have to wait :(
             */
            if (!isBillingAddressSame) {
                setBillingAddress();
            }
            /**
             * Ideally all these have to move to the cart as well.
             */
            setIsBillingAddressSame();
            /**
             * This needs to happen on the cart and we should call onSuccess
             * only when the mutation on cart is successful.
             *
             * And we need to check for errors.
             */
            setPaymentNonce(nonce);
            updateCCDetailsOnCart(nonce);
            if (onSuccess) {
                onSuccess(nonce);
            }
        },
        [
            onSuccess,
            setPaymentNonce,
            setBillingAddress,
            setIsBillingAddressSame,
            updateCCDetailsOnCart,
            isBillingAddressSame
        ]
    );

    const onPaymentError = useCallback(
        error => {
            if (onError) {
                onError(error);
            }
        },
        [onError]
    );

    const onPaymentReady = useCallback(() => {
        setDropinLoading(false);
        if (onReady) {
            onReady();
        }
    }, [onReady]);

    return {
        onPaymentError,
        onPaymentSuccess,
        onPaymentReady,
        isBillingAddressSame,
        countries,
        isDropinLoading
    };
};
