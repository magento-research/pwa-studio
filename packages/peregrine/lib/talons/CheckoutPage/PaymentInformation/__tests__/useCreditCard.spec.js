import React from 'react';
import { useQuery, useApolloClient, useMutation } from '@apollo/react-hooks';
import { useFormApi, useFormState } from 'informed';

import createTestInstance from '../../../../util/createTestInstance';
import { useCreditCard, mapAddressData } from '../useCreditCard';

/**
 * Mock Functions
 */

const getAllCountriesQuery = 'getAllCountriesQuery';
const getBillingAddressQuery = 'getBillingAddressQuery';
const getIsBillingAddressSameQuery = 'getIsBillingAddressSameQuery';
const getPaymentNonceQuery = 'getPaymentNonceQuery';
const getShippingAddressQuery = 'getShippingAddressQuery';

const setBillingAddressMutation = 'setBillingAddressMutation';
const setCreditCardDetailsOnCartMutation = 'setCreditCardDetailsOnCartMutation';

const billingAddress = {
    firstName: '',
    lastName: '',
    country: {
        code: ''
    },
    street: ['', ''],
    city: '',
    region: { code: '' },
    postalCode: '',
    phoneNumber: ''
};
const shippingAddress = {
    firstName: '',
    lastName: '',
    country: {
        code: ''
    },
    street: ['', ''],
    city: '',
    region: { code: '' },
    postalCode: '',
    phoneNumber: ''
};
const shippingAddressQueryResult = {
    data: {
        cart: {
            shippingAddresses: [
                {
                    __typename: 'Shipping Address',
                    ...shippingAddress
                }
            ]
        }
    }
};
const billingAddressQueryResult = {
    data: {
        cart: {
            billingAddress: {
                __typename: 'Billing Address',
                ...billingAddress
            }
        }
    }
};
const isBillingAddressSameQueryResult = {
    data: { cart: { isBillingAddressSame: false } }
};
const samplePaymentNonce = {
    details: {
        cardType: 'Visa',
        lastFour: '****',
        lastTwo: '**'
    },
    description: 'Visa card ending in **',
    type: 'braintree',
    nonce: '+++++++++++++++++'
};
const getAllCountries = jest.fn().mockReturnValue({ data: { countries: {} } });
const getBillingAddress = jest.fn().mockReturnValue(billingAddressQueryResult);
const getShippingAddress = jest
    .fn()
    .mockReturnValue(shippingAddressQueryResult);
const getIsBillingAddressSame = jest
    .fn()
    .mockReturnValue(isBillingAddressSameQueryResult);

const setBillingAddress = jest.fn();
const setCreditCardDetailsOnCart = jest.fn();

const setBillingAddressMutationResult = jest
    .fn()
    .mockReturnValue([setBillingAddress, {}]);
const setCreditCardDetailsOnCartMutationResult = jest
    .fn()
    .mockReturnValue([setCreditCardDetailsOnCart, {}]);

const writeQuery = jest.fn();

const queries = {
    getAllCountriesQuery,
    getBillingAddressQuery,
    getIsBillingAddressSameQuery,
    getPaymentNonceQuery,
    getShippingAddressQuery
};
const mutations = {
    setBillingAddressMutation,
    setCreditCardDetailsOnCartMutation
};

jest.mock('@apollo/react-hooks', () => {
    return {
        useQuery: jest.fn(),
        useApolloClient: jest.fn(),
        useMutation: jest.fn()
    };
});

jest.mock('../../../../context/cart', () => ({
    useCartContext: jest.fn().mockReturnValue([{ cartId: '123' }])
}));

jest.mock('informed', () => ({
    useFormState: jest.fn().mockReturnValue({
        values: {
            isBillingAddressSame: false,
            firstName: '',
            lastName: '',
            country: '',
            street1: '',
            street2: '',
            city: '',
            state: '',
            postalCode: '',
            phoneNumber: ''
        }
    }),
    useFormApi: jest.fn().mockReturnValue({
        setValue: () => {},
        setValues: () => {}
    })
}));

beforeAll(() => {
    useQuery.mockImplementation(query => {
        if (query === getAllCountriesQuery) {
            return getAllCountries();
        } else if (query === getBillingAddressQuery) {
            return getBillingAddress();
        } else if (query === getIsBillingAddressSameQuery) {
            return getIsBillingAddressSame();
        } else if (query === getShippingAddressQuery) {
            return getShippingAddress();
        } else {
            return { data: {} };
        }
    });

    useMutation.mockImplementation(mutation => {
        if (mutation === setBillingAddressMutation) {
            return setBillingAddressMutationResult();
        } else if (mutation === setCreditCardDetailsOnCartMutation) {
            return setCreditCardDetailsOnCartMutationResult();
        } else {
            return [jest.fn(), {}];
        }
    });

    useApolloClient.mockReturnValue({
        writeQuery
    });
});

const Component = props => {
    const talonProps = useCreditCard(props);

    return <i talonProps={talonProps} />;
};

const getTalonProps = props => {
    const tree = createTestInstance(<Component {...props} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    const update = newProps => {
        tree.update(<Component {...{ ...props, ...newProps }} />);

        return talonProps;
    };

    return { talonProps, tree, update };
};

/**
 * Tests
 */

test('Should return correct shape', () => {
    const { talonProps } = getTalonProps({
        queries,
        mutations,
        isHidden: false,
        onSuccess: () => {},
        onReady: () => {},
        onError: () => {}
    });

    expect(talonProps).toMatchSnapshot();
});

test('Shuold call onReady when payment is ready', () => {
    const onReady = jest.fn();
    const { talonProps } = getTalonProps({
        queries,
        mutations,
        isHidden: false,
        onSuccess: () => {},
        onReady,
        onError: () => {}
    });

    talonProps.onPaymentReady();

    expect(onReady).toHaveBeenCalled();
});

test('Shuold call onError when payment nonce generation errored out', () => {
    const error = 'payment error';
    const onError = jest.fn();
    const { talonProps } = getTalonProps({
        queries,
        mutations,
        isHidden: false,
        onSuccess: () => {},
        onError,
        onReady: () => {}
    });

    talonProps.onPaymentError(error);

    expect(onError).toHaveBeenCalledWith(error);
});

describe('Testing UI restoration', () => {
    const setValue = jest.fn();
    const setValues = jest.fn();

    beforeEach(() => {
        useFormApi.mockReturnValue({
            setValue,
            setValues
        });
    });

    test('UI fields should not be restored if payment method is hidden', () => {
        const { update } = getTalonProps({
            queries,
            mutations,
            isHidden: true,
            onSuccess: () => {},
            onReady: () => {},
            onError: () => {}
        });

        expect(setValue).not.toHaveBeenCalledWith(
            'isBillingAddressSame',
            false
        );
        expect(setValues).not.toHaveBeenCalledWith(billingAddress);
        expect(setValue).toHaveBeenCalledTimes(0);
        expect(setValues).toHaveBeenCalledTimes(0);

        update();

        expect(setValue).not.toHaveBeenCalledWith(
            'isBillingAddressSame',
            false
        );
        expect(setValues).not.toHaveBeenCalledWith(billingAddress);
        expect(setValue).toHaveBeenCalledTimes(0);
        expect(setValues).toHaveBeenCalledTimes(0);
    });

    test('UI fields should be restored if payment is not hidden and is ready, only once', () => {
        const { talonProps, update } = getTalonProps({
            queries,
            mutations,
            isHidden: false,
            onSuccess: () => {},
            onReady: () => {},
            onError: () => {}
        });

        /**
         * Payment is not ready yet, UI restoration should not happen.
         */

        expect(setValue).not.toHaveBeenCalledWith(
            'isBillingAddressSame',
            false
        );
        expect(setValues).not.toHaveBeenCalledWith(billingAddress);
        expect(setValue).toHaveBeenCalledTimes(0);
        expect(setValues).toHaveBeenCalledTimes(0);

        talonProps.onPaymentReady();

        /**
         * Updating the first time after payment ready. Should perform
         * UI restoration.
         */

        update();

        expect(setValue).toHaveBeenCalledWith('isBillingAddressSame', false);
        expect(setValues).toHaveBeenCalledWith(mapAddressData(billingAddress));
        expect(setValue).toHaveBeenCalledTimes(1);
        expect(setValues).toHaveBeenCalledTimes(1);

        /**
         * Updating again should not perform UI restoration.
         */

        update();

        expect(setValue).toHaveBeenCalledTimes(1);
        expect(setValues).toHaveBeenCalledTimes(1);
    });

    test('billingAddress should not be restored if it is null or undefined', () => {
        const billingAddressQueryResult = {
            data: {
                cart: {
                    billingAddress: null
                }
            }
        };

        getBillingAddress
            .mockReturnValueOnce(billingAddressQueryResult)
            .mockReturnValueOnce(billingAddressQueryResult);

        const { talonProps, update } = getTalonProps({
            queries,
            mutations,
            isHidden: false,
            onSuccess: () => {},
            onReady: () => {},
            onError: () => {}
        });

        talonProps.onPaymentReady();

        update();

        expect(setValue).toHaveBeenCalledWith('isBillingAddressSame', false);
        expect(setValues).not.toHaveBeenCalled();
    });

    test('UI fields should be restored everytime the payment method is shown after being hidden', () => {
        const { talonProps, update } = getTalonProps({
            queries,
            mutations,
            isHidden: false,
            onSuccess: () => {},
            onReady: () => {},
            onError: () => {}
        });

        /**
         * Payment method ready and component is not hidden.
         */

        talonProps.onPaymentReady();

        update();

        expect(setValue).toHaveBeenCalledWith('isBillingAddressSame', false);
        expect(setValues).toHaveBeenCalledWith(mapAddressData(billingAddress));
        expect(setValue).toHaveBeenCalledTimes(1);
        expect(setValues).toHaveBeenCalledTimes(1);

        /**
         * Setting component to hidden state. Hence
         * UI restoration should not happen.
         */

        update({
            isHidden: true
        });

        expect(setValue).toHaveBeenCalledTimes(1);
        expect(setValues).toHaveBeenCalledTimes(1);

        /**
         * Setting component to be visible. But since
         * dropin is not ready yet, UI restoration
         * should not happen.
         */

        update({
            isHidden: false
        });

        expect(setValue).toHaveBeenCalledTimes(1);
        expect(setValues).toHaveBeenCalledTimes(1);

        /**
         * Setting dropin to ready state.
         */

        talonProps.onPaymentReady();

        update();

        expect(setValue).toHaveBeenCalledTimes(2);
        expect(setValues).toHaveBeenCalledTimes(2);
    });
});

describe('Testing payment success workflow', () => {
    test('Should save payment payment nonce in apollo cache', () => {
        const paymentNonce = {
            details: 'payment details',
            description: 'payment made using the card ending in xxxx',
            type: 'visa',
            bin: {
                code: 'some bin data'
            }
        };
        const { talonProps } = getTalonProps({
            queries,
            mutations,
            isHidden: false,
            onSuccess: () => {},
            onReady: () => {},
            onError: () => {}
        });

        talonProps.onPaymentSuccess(paymentNonce);

        const paymentNonceSaveCall = writeQuery.mock.calls.filter(
            call => call[0].query === getPaymentNonceQuery
        )[0];

        expect(paymentNonceSaveCall[0].data.cart.paymentNonce).toMatchObject({
            details: 'payment details',
            description: 'payment made using the card ending in xxxx',
            type: 'visa'
        });
    });

    test('Should call setBillingAddressMutation mutation with billing address from UI if isBillingAddressSame is false', () => {
        const billingAddress = {
            firstName: 'test value',
            lastName: 'test value',
            country: 'test value',
            street1: 'test value',
            street2: 'test value',
            city: 'test value',
            state: 'test value',
            postalCode: 'test value',
            phoneNumber: 'test value'
        };
        useFormState.mockReturnValueOnce({
            values: {
                ...billingAddress,
                isBillingAddressSame: false
            }
        });

        const { talonProps } = getTalonProps({
            queries,
            mutations,
            isHidden: false,
            onSuccess: () => {},
            onReady: () => {},
            onError: () => {}
        });

        talonProps.onPaymentSuccess(samplePaymentNonce);

        expect(setBillingAddress).toBeCalledWith({
            variables: {
                ...billingAddress,
                sameAsShipping: false,
                cartId: '123'
            }
        });
    });

    test('Should call setBillingAddressMutation mutation with shipping address from UI if isBillingAddressSame is true', () => {
        const shippingAddress = {
            firstName: 'test value',
            lastName: 'test value',
            country: {
                code: 'test value'
            },
            street: ['test value', 'test value'],
            city: 'test value',
            region: { code: 'test value' },
            postalCode: 'test value',
            phoneNumber: 'test value'
        };
        getShippingAddress.mockReturnValueOnce({
            data: {
                cart: {
                    shippingAddresses: [
                        {
                            __typename: 'Shipping Address',
                            ...shippingAddress
                        }
                    ]
                }
            }
        });
        useFormState.mockReturnValueOnce({
            values: {
                isBillingAddressSame: true
            }
        });

        const { talonProps } = getTalonProps({
            queries,
            mutations,
            isHidden: false,
            onSuccess: () => {},
            onReady: () => {},
            onError: () => {}
        });

        talonProps.onPaymentSuccess(samplePaymentNonce);

        expect(setBillingAddress).toBeCalledWith({
            variables: {
                ...mapAddressData(shippingAddress),
                sameAsShipping: true,
                cartId: '123'
            }
        });
    });

    test('Should save isBillingAddressSame in apollo cache', () => {
        useFormState.mockReturnValueOnce({
            values: {
                isBillingAddressSame: true
            }
        });

        const { talonProps } = getTalonProps({
            queries,
            mutations,
            isHidden: false,
            onSuccess: () => {},
            onReady: () => {},
            onError: () => {}
        });

        talonProps.onPaymentSuccess(samplePaymentNonce);

        const isBillingAddressSameSaveCall = writeQuery.mock.calls.filter(
            call => call[0].query === getIsBillingAddressSameQuery
        )[0];

        expect(
            isBillingAddressSameSaveCall[0].data.cart.isBillingAddressSame
        ).toBeTruthy();
    });
});

test('Should call setCreditCardDetailsOnCartMutation on payment success', () => {
    const { talonProps } = getTalonProps({
        queries,
        mutations,
        isHidden: false,
        onSuccess: () => {},
        onReady: () => {},
        onError: () => {}
    });

    talonProps.onPaymentSuccess(samplePaymentNonce);

    expect(setCreditCardDetailsOnCart).toHaveBeenCalledWith({
        variables: {
            cartId: '123',
            paymentMethod: 'braintree',
            paymentNonce: samplePaymentNonce.nonce
        }
    });
});

test('Should call onSuccess if setCreditCardDetailsOnCartMutation is successful', () => {
    setCreditCardDetailsOnCartMutationResult.mockReturnValueOnce([
        jest.fn(),
        {
            called: true,
            loading: false,
            error: null
        }
    ]);

    const onSuccess = jest.fn();
    const { talonProps } = getTalonProps({
        queries,
        mutations,
        isHidden: false,
        onSuccess,
        onReady: () => {},
        onError: () => {}
    });

    talonProps.onPaymentSuccess(samplePaymentNonce);

    expect(onSuccess).toHaveBeenCalled();
});

test('Should not call onSuccess if setCreditCardDetailsOnCartMutation is not successful', () => {
    setCreditCardDetailsOnCartMutationResult.mockReturnValueOnce([
        jest.fn(),
        {
            called: true,
            loading: false,
            error: ['some error']
        }
    ]);

    const onSuccess = jest.fn();
    const { talonProps } = getTalonProps({
        queries,
        mutations,
        isHidden: false,
        onSuccess,
        onReady: () => {},
        onError: () => {}
    });

    talonProps.onPaymentSuccess(samplePaymentNonce);

    expect(onSuccess).not.toHaveBeenCalled();
});
