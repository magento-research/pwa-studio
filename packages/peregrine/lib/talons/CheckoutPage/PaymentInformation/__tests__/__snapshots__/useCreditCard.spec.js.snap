// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return correct shape 1`] = `
Object {
  "countries": Object {},
  "errors": Array [],
  "isBillingAddressSame": false,
  "isDropinLoading": true,
  "onPaymentError": [Function],
  "onPaymentReady": [Function],
  "onPaymentSuccess": [Function],
}
`;
